write-fonts/src/graph/splitting.rs

```diff
@@ -188,6 +186,247 @@ fn split_off_ppf1(graph: &mut Graph, subtable: ObjectId, start: u16, end: u16) -
+        // Please remind me to open an issue to investigate size measurement
+        // more thoroughly. In particular, look into why we just take subgraph
+        // size for pairpos1 (ignoring duplicates) but try to account for duplicates
+        // in pairpos2?
+        // tracked at <https://github.com/googlefonts/fontations/issues/601>
+        if has_device_tables {
+            for class2rec in class1rec.unwrap().class2_records.iter() {
+                let class2rec = class2rec.as_ref().unwrap();
+                accumulated_delta += size_of_value_record_children(
+                    &class2rec.value_record1,
+                    graph,
+                    &data.offsets,
+                    &mut next_device_offset,
+                    &mut visited,
+                );
+                accumulated_delta += size_of_value_record_children(
+                    &class2rec.value_record2,
+                    graph,
+                    &data.offsets,
+                    &mut next_device_offset,
+                    &mut visited,
+                );
+            }
+        }
+
+        accumulated += accumulated_delta;
+        let largest_obj = coverage_size.max(class_def_1_size).max(class_def2_size);
+        let total = accumulated + coverage_size + class_def_1_size + class_def2_size
+            // largest obj packs last and can overflow
+            - largest_obj;
+
+        if total > MAX_TABLE_SIZE {
+            split_points.push(idx);
+            // split does not include this class, so add it for the next iteration
+            accumulated = BASE_SIZE + accumulated_delta;
+            coverage_size = 4 + estimator.increment_coverage_size(idx as _);
+            class_def_1_size = 4 + estimator.increment_class_def_size(idx as _);
+            visited.clear();
+        }
+    }
+
+    log::debug!("identified {} split points", split_points.len());
+    if split_points.is_empty() {
+        return None;
+    }
+
+    split_points.push(pp2.class1_count() as usize - 1);
+    // now we have a list of split points, and just need to do the splitting.
+    // note: harfbuzz does a thing here with a context type and an 'actuate_splits'
+    // method.
+
+    let mut new_subtables = Vec::new();
+    let mut prev_subtable_start = 0;
+    let mut next_device_offset = 3; // after coverage & two class defs
+    for subtable_start in split_points {
+        let subtable_end = subtable_start - 1;
+        let (new_subtable, offsets_used) = split_off_ppf2(
+            graph,
+            subtable,
+            prev_subtable_start,
+            subtable_end,
+            next_device_offset,
+        );
+        prev_subtable_start = subtable_start;
+        next_device_offset += offsets_used;
+        new_subtables.push(graph.add_object(new_subtable));
+    }
+
+    Some(new_subtables)
+}
+
+// returns the new table, + the number of non-null device offsets encountered.
+fn split_off_ppf2(
```

Comment:
This function seems to be a combination of some `PairPosFormat2::split_context_t` methods and `actuate_subtable_split`. I like the simplifications you've managed here but future readers might appreciate a couple comments with links to relevant portions of the HB code.

Reply:
yea there are some specific details of the harfbuzz implementation I haven't followed here; in particular harfbuzz shrinks the original table and retains it, whereas I just generate a full set of new tables. There are a few comments scattered about but I can try to rationalize that.